### ISSUE TITLE
After restarting Cinnamon, updates cannot be applied anymore

### ISSUE THREAD

**Issue by IzzySoft** [linuxmint/mintupdate#719]
**Describe the bug**
I had to restart Cinnamon again, [because](https://github.com/linuxmint/cinnamon/issues/6850). New updates arrived and were indicated in the tray. Clicking the tray, updater window opens – but clicking "Install Updates" just greys out the window, nothing else happens; no password prompt, no updates. Other windows can be opened, e.g. kernels – but no actions are possible there either (i.e. one cannot remove or install a kernel).

* Completely exiting mintUpdate (right-click on the tray icon, Exit) and starting via the menu didn't help
* …starting via <kbd>Alt</kbd>-<kbd>F2</kbd> didn't help
* …starting from a terminal didn't help, and didn't give any clues from terminal output either

**Screenshots**
n/a

**To Reproduce**
1. I've opened a console via <kbd>Ctrl</kbd>-<kbd>Alt</kbd>-<kbd>F1</kbd> and sent a `HUP` to the `cinnamon --replace` process (see linked bug for details)
2. Returned to the graphical desktop via <kbd>Alt</kbd>-<kbd>F1</kbd>
3. Confirmed reloading Cinnamon (was in fallback mode)
4. Opened mintUpdate (see above)

**Expected behavior**
mintUpdate fully operational – especially prompting for the password and perform the requested updates.

**Distribution:**
- [x] Linux Mint 20.2
- [ ] LMDE
- [ ] Other (please specify)

**Software version:**
5.7.8

**Logs:**
Copy & paste the log output from Update Manager > View > Information here.
==> not possible. At that stage, the window is grayed out and menus cannot be accessed anymore.

Found the logs in `/tmp/mintUpdate` however:

```
01.29@22:13 ++ Install requested by user
01.29@22:13 ++ Will install firefox-locale-en
01.29@22:13 ++ Will install firefox
01.29@22:13 ++ Will install dino
01.29@22:13 ++ Ready to launch synaptic
01.29@22:14 ++ Exiting - requested by user
```

So you see it says "Ready to launch synaptic", but has no update actions performed. A little more than a minute later, I closed the updater via the tray.

**Crash report:**
If you are reporting a crash, please run `mintupdate` from a terminal window, reproduce the crash and copy & paste any output from the terminal window here.
==> no crash, it simply hangs. No additional console output either. Output right at start:

```
/usr/lib/linuxmint/mintUpdate/mintUpdate.py:1274: DeprecationWarning: Gdk.threads_init is deprecated
  Gdk.threads_init()
/usr/lib/linuxmint/mintUpdate/mintUpdate.py:1422: DeprecationWarning: Gtk.ImageMenuItem.set_image is deprecated
  menuItem3.set_image(image)
/usr/lib/linuxmint/mintUpdate/mintUpdate.py:722: DeprecationWarning: Gdk.threads_enter is deprecated
  Gdk.threads_enter()
/usr/lib/linuxmint/mintUpdate/mintUpdate.py:729: DeprecationWarning: Gdk.threads_leave is deprecated
  Gdk.threads_leave()
```

**Additional context**
Starting `synaptic` from the command line (with or without `sudo`) works fine, so it can be invoked. Starting it via MintMenu however fails the same way it does from mintUpdate. Looking up the desktop file I can see the command executed is `synaptic-pkexec` – trying that from a terminal simply hangs. No hints from syslog. But checking the process list I see multiple `synaptic-pkexec` processes are still running, at least one of them (PID 475298) predating the restart of Cinnamon:

```
USER         PID    PPID S CPU %MEM   RSS    VSZ  START     ELAPSED   TIME CMD
pete      475934  405296 S   -  0.0   604   2616  00:12       04:55   0:00 /bin/sh /usr/bin/synaptic-pkexec
root      475298  475134 S   -  0.0  6948 159184  00:03       14:06   0:00 pkexec /usr/sbin/synaptic --hide-main-window --non-interactive --parent-window-id 142606349 -o Synaptic::closeZvt=true --set-selections-file /tmp/tmpc_itd13s
root      475936  475934 S   -  0.0  7344 159184  00:12       04:55   0:00 pkexec /usr/sbin/synaptic
```

However, killing those processes does not change the behavior, and no new processes show up either, nor can `synaptic-pkexec` be started via menu or from a terminal. `/root/.synaptic/log` has no recent entries (only logs from before the last boot about a week ago). The only files there matching the time of the issue are an empty `options` and a `synaptic-mintupdate-workaround.conf` file in `/root/.synaptic` (the lock files there are 2 months old, around the time of the fresh install of this machine – and yes, I've applied multiple updates after that via mintUpdate, including kernel updates and removals).

A boot will probably solve that, but we're not on Windows here, are we? :wink: That could only be named a temporary work-around at best.

Possibly related:

* [Synaptic Package Manager Hangs Up](https://forums.linuxmint.com/viewtopic.php?f=47&t=366404&sid=df8f364291f64d9f425e167b05bd0926) (Mint forum, a thread that started a week ago and ended unsuccessful yesterday)
* [<solved> mintUpdate does not install stuff](https://forums.linuxmint.com/viewtopic.php?f=47&t=194290&sid=df8f364291f64d9f425e167b05bd0926) (Mint forum again): a user reports that `sudo chmod u+s /usr/bin/pkexec` solved the issue for them. However, at least on my system the file is already u+s (4755/-rwsr-xr-x)

**Locale:**
If the problem is with text not getting displayed correctly, paste the output of the terminal command `locale`.
==> No.

**Comment by IzzySoft** [comment]
PS: I ran an `strace synaptic-pkexec`. These are the last lines:

```
stat("/home/pete/bin/pkexec", 0x7ffe5ee354f0) = -1 ENOENT (No such file or directory)
stat("/usr/local/sbin/pkexec", 0x7ffe5ee354f0) = -1 ENOENT (No such file or directory)
stat("/usr/local/bin/pkexec", 0x7ffe5ee354f0) = -1 ENOENT (No such file or directory)
stat("/usr/sbin/pkexec", 0x7ffe5ee354f0) = -1 ENOENT (No such file or directory)
stat("/usr/bin/pkexec", {st_mode=S_IFREG|S_ISUID|0755, st_size=31032, ...}) = 0
clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7f2fe1fea850) = 477842
wait4(-1, 
```

And there it sits, waiting for Godot (or whatever it is waiting for). Could it have anything to do with the recent polkit update? As Timeshift cannot be started either from MintMenu (it uses polkit as well to obtain elevated privileges), this might well be the case. So that this issue seemed to start after a Cinnamon restart might be pure coincidence. As with synaptic, I can start Timeshift fine via `sudo` from a terminal (`sudo timeshift-gtk`).

Possibly related:

* [LM 20.3: Update problem with "pkexec"](https://forums.linuxmint.com/viewtopic.php?f=47&t=366726&sid=a2dd501a981f0a5bc9de0ab5f19f96a1) (I've already applied updates via `apt upgrade` twice meanwhile as mintUpdate won't let me, so this is not solving my issue)

To confirm my suspicion, I've installed `ssh-askpass`, created a file `/usr/local/bin/pkexec` with the content bellow, and made it executable for all. Et voila: I can start Timeshift again from MintMenu, kernels can be removed again via mintUpdate…

```bash
#!/bin/bash
SUDO_ASKPASS=/usr/bin/ssh-askpass sudo -A $@
```

**TL;DR:** Polkit seems to be broken, causing these issues. A work-around is what the last paragraph describes. But am I the only one affected this way? After the next kernel update (and its following reboot) I'll check if the work-around is still needed, or a reboot solves it. Feel free to ping me should I forget to report back within a week.
